src/utils.cpp

```diff
@@ -827,11 +727,14 @@ std::tuple<std::string, std::string> get_kernel_dirs(
     kobj = "";
   }
   if (ksrc.empty() && kobj.empty()) {
-    if (unpack_kheaders) {
-      const auto kheaders_tar_xz_path = unpack_kheaders_tar_xz(utsname);
-      if (kheaders_tar_xz_path.size() > 0)
-        return std::make_tuple(kheaders_tar_xz_path, kheaders_tar_xz_path);
-    }
+    LOG(WARNING) << "Could not find kernel headers in " << ksrc << " or "
+                 << kobj
+                 << ". To specify a particular path to kernel headers, set the "
+                    "env variables BPFTRACE_KERNEL_SOURCE and, optionally, "
+                    "BPFTRACE_KERNEL_BUILD if the kernel was built in a "
+                    "different directory than its source. To create kernel "
+                    "headers run 'modprobe kheaders', which will create a tar "
+                    "file at /sys/kernel/kheaders.tar.xz";
```

Comment:
I imagine that in most cases bpftrace doesn't need kheaders (can use BTF data instead); would it make sense to only print this warning if kernel headers end up being required?

A couple of cases come to mind:

- Bpftrace script unconditionally includes one or more kernel headers (e.g. `#include <linux/sched.h>` without gating on `BPFTRACE_HAVE_BTF`)
- Conflicts between user-defined and BTF types

Related: [commit 91db77c5](https://github.com/bpftrace/bpftrace/commit/91db77c5707c30346035330ea4fb719821254161)

cc @danobi

Reply:
This is a good point. I agree that we should not print the warning by default as there's a number of situations where kheaders are not necessary.

Reply:
How about as a stop-gap, I just reinstate the `!bpftrace.feature_->has_btf()` check as a conditional to show the warning?

Reply:
I may remember it wrong (it's been a while since I wrote that code) but IMHO kheaders are needed iff ClangParser runs so we should use the same condition as in [clang_parser.cpp:663](https://github.com/bpftrace/bpftrace/blob/master/src/clang_parser.cpp#L663):
```cpp
  if (program->c_definitions.empty() && bpftrace.btf_set_.empty())
```
It'd be nice if we could reuse it.

An alternative would be to issue the warning from ClangParser when it fails but that may be even more complicated.

Reply:
I moved a few things around but I'm using this same check before even calling `get_kernel_dirs` so the warning should only be shown in that case.